src/libmeasurement_kit/common/logger.cpp

```diff
@@ -188,6 +195,34 @@ class DefaultLogger : public Logger, public NonCopyable, public NonMovable {
                 /* Suppress */;
             }
         }
+        assert(!!s);
+        // Note that the mutex is recursive
+        // TODO(bassosimone): improve the API to allow emitting more context
+        emit_event_ex("status.progress", {
+            {"percentage", prog},
+            {"message", s},
+        });
+    }
+
+    void emit_event_ex(
+            const std::string &key, nlohmann::json &&value) override {
+        if (!value.is_object()) {
+            warn("wrong value for key: %s", key.c_str());
+            assert(false);
+        }
+        if (handlers_.count(key) <= 0) {
+            return;
+        }
+        nlohmann::json event{
+            {"key", key},
+            {"value", std::move(value)}
+        };
+        std::unique_lock<std::recursive_mutex> _{mutex_};
+        // TODO(bassosimone): other logging functions filter all the
+        // exceptions. We cannot change this behavior until that is part
+        // of our public API. But here we deliberately choose not to do
+        // any exception handling. The callee must behave.
+        handlers_.at(key)(std::move(event));
```

Comment:
`.at()` throws when the key is missing. Is that the desired behavior?

Reply:
Throwing is not desired, but the above `handles_.count(key) <= 0` check should guarantee that, if we arrive down this line, there is a registered handler, right?

Reply:
Yup! I can't believe I missed that.